### ISSUE TITLE
[relay-hooks 3.0.0] useLazyLoadQuery & support relay-runtime v8.0.0

### ISSUE THREAD

**Issue by morrys** [relay-tools/relay-hooks#53]
branch: https://github.com/relay-tools/relay-hooks/tree/suspense-v8

* [X] create useLazyLoadQuery https://github.com/relay-tools/relay-hooks/commit/98d0f27dd71d8935b22ac6ae4d5ab383e3ab5d49
* [X] improved queryFetcher & useQueryFetcher to handle both useQuery (no suspense) and useLazyLoadQuery (with suspense) https://github.com/relay-tools/relay-hooks/commit/75bccc96add5a5480b120dcfeae3a32cf614fbad
* [X] support relay-runtime v8.0.0 https://github.com/relay-tools/relay-hooks/commit/1cc3560874db728f0658f5bebcf590e898eb7543
* [X] added tests from relay-experimental ( one test removed,  ) https://github.com/relay-tools/relay-hooks/commit/5cdc866e8d47b71948c66013762669356920e23c
* [X] evaluate an implementation of temporary retain https://github.com/relay-tools/relay-hooks/commit/2fbe61d8f022c93b8eb4e733c5e296969f46e40f
* [X] integration tests
* [X] created [nextjs example with ssr & useLazyLoadQuery](https://github.com/relay-tools/relay-hooks/tree/suspense-v8/examples/suspense/nextjs-ssr)
* [X] release 3.0.0-rc.1
* [X] created useRefetchable https://github.com/relay-tools/relay-hooks/commit/c2f671b492075d6c5a577a5781e1055e7924742f
* [X] added tests for useRefetchable
* [X] release 3.0.0-rc.2

**Comment by morrys** [comment]
taking advantage of the major release I would like to evaluate the change of return type of useQuery & useLazyLoadQuery

**now:**
```ts
export interface RenderProps<T extends OperationType> {
    error: Error;
    props: T['response'];
    retry: (_cacheConfigOverride?: CacheConfig) => void;
    cached?: boolean;
}
```
const {error, cached, props, retry} = useQuery(query, variables);


**after:**
```ts
export type QueryResult<T extends OperationType> = [T['response'], RenderProps];

export interface RenderProps {
    error: Error;
    retry: (_cacheConfigOverride?: CacheConfig) => void;
    cached?: boolean;
}
```
const [data, {error, cached, retry}] = useQuery(query, variables);

@dminkovsky @jazblueful @sibelius  what do you think?

**Comment by jazblueful** [comment]
Although second option is probably more 'reacty', I'm using the first pattern for return type of all of my custom hooks so that is my personal preference.

**Comment by morrys** [comment]
@jazblueful  thanks for your opinion :)

I'm currently evaluating if I can integrate @refetchable in version 3.0.0 in the meantime let's see if someone else is of a different opinion, otherwise I release the current implementation :)

**Comment by morrys** [comment]
hi @jazblueful, today i released relay-hooks 3.0.0-rc.2 with useRefetchable:

```ts
function useRefetchable<TKey extends KeyType>(
    fragmentInput: GraphQLTaggedNode,
    fragmentRef: TKey,
): [$Call<KeyReturnType<TKey>>, RefetchableFunction];
function useRefetchable<TKey extends KeyType>(
    fragmentInput: GraphQLTaggedNode,
    fragmentRef: TKey | null,
): [$Call<KeyReturnType<TKey>> | null, RefetchableFunction];
function useRefetchable<TKey extends ArrayKeyType>(
    fragmentInput: GraphQLTaggedNode,
    fragmentRef: TKey,
): [ReadonlyArray<$Call<ArrayKeyReturnType<TKey>>>, RefetchableFunction];
function useRefetchable<TKey extends ArrayKeyType>(
    fragmentInput: GraphQLTaggedNode,
    fragmentRef: TKey | null,
): [ReadonlyArray<$Call<ArrayKeyReturnType<TKey>>> | null, RefetchableFunction]

export type RefetchableFunction = (
    refetchVariables: Variables | ((fragmentVariables: Variables) => Variables),
    options?: {
        renderVariables?: Variables;
        observerOrCallback?: ObserverOrCallback;
        refetchOptions?: RefetchOptions;
    },
) => Disposable;


const fragmentSpec = graphql`
  fragment TodoApp_user on Query
    @refetchable(queryName: "UserFragmentRefetchQuery") {
    user(id: $userId) {
      id
      userId
      totalCount
      completedCount
      ...TodoListFooter_user
      ...TodoList_user
    }
  }
`;
const [{user}, refetch] = useRefetchable(fragmentSpec, props.query);

...
<StyledButton
            onClick={() => {
              refetch({
                userId: user.userId === 'me' ? 'you' : 'me',
              });
            }}>
            Change User
          </StyledButton>
...
```

Do you have the opportunity to try this version to confirm compatibility with relay v8?

thanks

**Comment by jazblueful** [comment]
Hi @morrys,

Good job! I updated all the relay related dependencies including relay hooks to the latest and so far everything is working as expected :)

I'm not using useRefetchable anywhere yet though. That being said I think I have fallen way behind on understanding what functions are most appropriate to use regarding relay and relay-hooks or if relay-hooks is even necessary considering relay now has (some) support for hooks.

**Comment by morrys** [comment]
Hi @jazblueful,
I'm sorry these days I've been very busy: /

Meanwhile, thank you for making the test, as soon as possible I will release version 3.0.0.

As for relay-hooks I can understand your doubts but as long as suspense and the official relay hooks are in an experimental phase I will certainly continue to update this repository.
For the moment I think it is a very stable library without constraints dictated by fb :) We will see later how things evolve

**Comment by morrys** [comment]
released relay-hooks 3.0.0 https://github.com/relay-tools/relay-hooks/releases/tag/v3.0.0
